cpp/include/graph.hpp

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+namespace cugraph {
+namespace experimental {
+
+typedef enum prop_type{PROP_UNDEF, PROP_FALSE, PROP_TRUE} PropType;
+
+struct GraphProperties {
+  bool directed;
+  bool weighted;
+  bool multigraph;
+  bool bipartite;
+  bool tree;
+  PropType has_negative_edges;
+  GraphProperties() : directed(false), weighted(false), multigraph(false), bipartite(false), tree(false), has_negative_edges(PROP_UNDEF){}
+};
+
+/**
+ * @Synopsis    Base class graphs, all but vertices and edges
+ *
+ * @tparam VT   Type of vertex (defaults to int)
+ * @tparam WT   Type of weight (defaults to float)
+ */
+template <typename WT>
+class GraphBase {
+public:
+  WT const *edge_data;     ///< edge weight
+
+  GraphProperties          prop;
+
+  size_t                   number_of_vertices;
+  size_t                   number_of_edges;
+
+  GraphBase(WT const *edge_data_, size_t number_of_vertices_, size_t number_of_edges_):
+    edge_data(edge_data_),
+    prop(),
+    number_of_vertices(number_of_vertices_),
+    number_of_edges(number_of_edges_)
+  {}
+};
+
+/**
+ * @Synopsis    A graph stored in COO (COOrdinate) format.
+ *
+ * @tparam VT   Type of vertex (defaults to int)
+ * @tparam WT   Type of weight (defaults to float)
+ */
+template <typename VT = int, typename WT = float>
+class GraphCOO: public GraphBase<WT> {
+public:
+  VT const *src_indices;   ///< rowInd
+  VT const *dst_indices;   ///< colInd
```

Comment:
And it's safe to always initialize class member variables.

VT const *src_indices{nullptr};
VT const *dst_indices{nullptr};